Probe libnuma.so.1.0.0 before its symlinks

am11 opened the pull request:
On Linux, libnuma.so{.1} are symlinks to libnuma.so.1.0.0.

When installed from default package manager; Ubuntu (libnuma-dev) and
Alpine (numactl-dev), found these two symlinks:
`{libnuma.so,libnuma.so.1} -> libnuma.so.1.0.0`

On Fedora (numactl-libs):
`libnuma.so.1 -> libnuma.so.1.0.0`
(there is no version-less variant)

PR adjusts probing fallback flow based on this info.

am11 commented:
cc @janvorli, this saves us from copying symlink in sandbox manifest.

janvorli reviewed (approved):
LGTM, thank you!

tmds commented:
@am11 @janvorli fyi. My experience with Fedora packages is that library packages have versioned .so files, and development packages include the non-versioned symlinks. Fedora has the versionless numa symlink in `numactl-devel`. Production machines won't have development packages, and will be missing the versionless symlinks.

janvorli commented:
@tmds I know. And this change actually makes the probing better - the versionless one is now the last one tried instead of the first one.

tmds commented on src/coreclr/gc/unix/gcenv.unix.cpp:
```diff
@@ -284,10 +284,14 @@ void NUMASupportInitialize()
     }
 
-    g_numaHandle = dlopen("libnuma.so", RTLD_LAZY);
+    g_numaHandle = dlopen("libnuma.so.1.0.0", RTLD_LAZY);
```
I think `libnuma.so.1.0.0` could be left out. There is no need to probe with a specific minor and patch version.

janvorli replied:
Adding that was actually the point of this PR. @am11 made this change to make .NET work on a system when there is no symlink libnuma.so.1, just the libnuma.so.1.0.0

tmds replied:
@am11 when I check nothing seems missing on Alpine:

```$ podman run -ti alpine /bin/sh -c 'apk update && apk add numactl && ls -lah /usr/lib/libnuma*'
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
v3.13.1-98-g655b7d332a [https://dl-cdn.alpinelinux.org/alpine/v3.13/main]
v3.13.1-99-ge77be6b709 [https://dl-cdn.alpinelinux.org/alpine/v3.13/community]
OK: 13878 distinct packages available
(1/1) Installing numactl (2.0.14-r0)
OK: 6 MiB in 15 packages
lrwxrwxrwx    1 root     root          16 Feb 11 06:30 /usr/lib/libnuma.so.1 -> libnuma.so.1.0.0
-rwxr-xr-x    1 root     root       46.2K Dec 14 14:11 /usr/lib/libnuma.so.1.0.0
```

am11 replied:
Hmm, that's what I have also found and shown in the description above, or maybe I missed something?

SO version 1.0.0 (as opposed to the package version 2.0.14) of libnuma hasn't changed since the beginning (~10y), so it doesn't affect anything especially with the existing probing paths intact.

This PR is a minor improvement (on top of https://github.com/dotnet/coreclr/pull/12556) mainly for the sandbox environments, where we don't have/want to carry around a symlink in addition to the binary.

tmds replied:
`libnuma.so.1` is present on all Linux systems after installing the appropriate package.
I think we should leave out the `libnuma.so.1.0.0`: there is no need for probing for a specific minor.patch version.
Does that make sense?

am11 replied:
Nope. Adding the additional probe was the whole purpose of this PR for the reasons described above. :)

janvorli replied:
@am11, I think we can swap the order of libnuma.so.1 and libnuma.so.1.0.0 here. That would address the concern that the library version might change in the future on some systems and it would still work for your case. Does it sound reasonable?

tmds replied:
@am11 from what is said, I don't understand what platform is missing `libnuma.so.1` and therefore requires the specific `libnuma.so.1.0.0` probe.

The version numbers on these `.so` files are "semantic versioning". When the major increases that indicates a breaking change. We're using the v1 API, so `libnuma.so.1` is the preferable file to use.

am11 committed:
Probe libnuma.so.1.0.0 before its symlinks
On Linux, libnuma.so{.1} are symlinks to libnuma.so.1.0.0.

When installed from default package manager; Ubuntu (libnuma-dev) and
Alpine (numactl-dev), found these two symlinks:
`{libnuma.so,libnuma.so.1} -> libnuma.so.1.0.0`

On Fedora (numactl-libs):
`libnuma.so.1 -> libnuma.so.1.0.0`
(there is no version-less variant)

PR adjusts probing fallback flow based on this info.

am11 commented:
win-x64 installer leg failure is #48169.
